doc/internals/01_directory_structure.md

```diff
@@ -0,0 +1,207 @@
+## pgagroal directory structure
+
+This document takes one through the pgagroal directory structure from a high level
+it is not meant to be a tutorial for end users, but for developers who are trying
+to contribute to pgagroal
+
+- Last Update time: 25/04/2023 (DD/MM/YYYY)
+
+### Source code
+
+Source code is found in the `/src` directory.
+
+Headers are placed in `src/include/*.h` while the corresponding source files are in
+`libpgagroal`
+
+### 1. `configuration.h /configuration.c`
+
+Contains configuration parsing code
+
+`configuration.h` exposes the following configuration parsers
+
+- `pgagroal_read_configuration` : Read main configuration
+- `pgagroal_read_hba_configuration`: Read HBA configuration
+- `pgagroal_read_limit_configuration` : Read limit configuration
+- `pgagroal_read_users_configuration`: Read users configuration
+- `pgagroal_read_frontend_users_configuration`: Read frontend users configuration
+- `pgagroal_read_admins_configuration`: Read admin configuration
+- `pgagroal_read_superuser_configuration`: Read superuser from a file.
+- `pgagroal_reload_configuration`: Reload configuration
+- `pgagroal_write_config_value`: Write configuration values, supporting `server`,`hba`,`limit` and `main` config files.
+
+The configuration keys cannot exceed `MISC_LENGTH` ( which is currently defined as 128),
+if such keys exist they are treated as errors.
+
+### 2. `logging.h/logging.c`
+
+Contain logging configurations
+
+Verbosity of logging is configured by setting it in `pgagroal.conf` dynamically
+there are six levels, from `trace` to `fatal`.
+
+It also exposes the ability to do log rotation
+
+The following indicate levels a config file generates given a level
+
+- `debug5-debug0` - pgagroal supports multiple debug levels, each increasing verbosity, `debug5`  is the highest
+  level , which activates the trace level of logging,
+
+- `info`: Enables info level logging
+- `warn`: Enables warn level logging
+- `error`: Enables error level logging
+- `fatal`: Enables fatal level logging
+
+To configure logging you set the appropriate level in the config file (`pgagroal.conf`)
+
+```toml
+[pgagroal]
+# set level here, can be 
+# debug5,debug0,info,warn,error,fatal
+log_level = info
+```
+
+There are no quotations in between configurations.
+
+### 3. `management.h/management.c`
+
+This contains code providing boilerplate for reading and writing to sockets
+e.g error recovery and connection management.
+
+### 4. `memory.h/memory.c`
+
+These contain routines to manage two variables `message` and `data`.
+
+The stated variables are global and the functions in `memory` help allocate, set and delete
+the above memory.
```

Comment:
This is the code that make pgagroal don't allocate memory at run-time

Reply:
Nice, I didn't realize that, this seems like a nice feature to mention in the readme